File: telemetry/aitf/sdk/python/src/aitf/ocsf/event_classes.py
"""AITF OCSF AI Event Classes (class-reuse model).

Defines the ten AI event types, each mapped onto its reused OCSF class rather
than a bespoke category. Based on event classes from the AITelemetry project,
extended for AITF with MCP, Skills, ModelOps, Asset Inventory, and enhanced
agent support.

Every ``class_uid`` here is a **released OCSF v1.9.0 class**. AITF's earlier
Category 7 (7001-7010) and the later provisional ``ai`` category (9001-9003)
have both been retired: v1.9.0 solved agent attribution by landing ``ai_agent``
and ``delegation`` on the ``ai_operation`` profile and attaching that profile to
the ``system``, ``network``, ``application`` and ``iam`` base classes, and it
ships no category above uid 8. See ``schema.LEGACY_AI_CLASS_UIDS``.
"""

from __future__ import annotations

from typing import Any

from pydantic import Field

from aitf.ocsf.schema import (
    AIBaseEvent,
    AICostInfo,
    AILatencyMetrics,
    AIModelInfo,
    AISecurityFinding,
    AITeamInfo,
    AITokenUsage,
    OCSFAgentMessage,
    OCSFCategoryUID,
    OCSFClassUID,
)


class AIModelInferenceEvent(AIBaseEvent):
    """AI model inference — reuses OCSF API Activity (6003).

    The model call is an API operation; AI specifics ride on the
    ``ai_operation`` profile (``ai_agent``, ``ai_model``).
    """
    category_uid: int = OCSFCategoryUID.APPLICATION
    class_uid: int = OCSFClassUID.API_ACTIVITY
    model: AIModelInfo
    token_usage: AITokenUsage = Field(default_factory=AITokenUsage)
    latency: AILatencyMetrics | None = None
    request_content: str | None = None
    response_content: str | None = None
    streaming: bool = False
    tools_provided: int = 0
    finish_reason: str = "stop"
    cost: AICostInfo | None = None
    error: dict[str, Any] | None = None


class AIAgentActivityEvent(AIBaseEvent):
    """Agent lifecycle and reasoning steps — reuses OCSF API Activity (6003).

    An agent step *is* an operation, so it maps onto API Activity and carries
    attribution on the ``ai_operation`` profile (``ai_agent``, ``delegation``),
    which OCSF v1.9.0 attached to the ``application`` base class.

    Previously emitted as provisional ``agent_activity`` (9001) in a proposed
    ``ai`` category. That category was never ratified — OCSF v1.9.0 ships no
    category above uid 8 — so it has been retired. Deployments that model agent
    *session* start/stop separately from reasoning steps MAY emit those two
    activities as Application Lifecycle (6002) instead; both are category 6 and
    both carry the same ``ai_operation`` attribution.
    """
    category_uid: int = OCSFCategoryUID.APPLICATION
    class_uid: int = OCSFClassUID.API_ACTIVITY
    agent_name: str
    agent_id: str
    agent_type: str = "autonomous"
    framework: str | None = None
    session_id: str
    step_type: str | None = None
    step_index: int | None = None
    thought: str | None = None
    action: str | None = None
    observation: str | None = None
    delegation_target: str | None = None
    team_info: AITeamInfo | None = None


class AIAgentCommunicationEvent(AIBaseEvent):
    """Agent-to-agent communication (A2A / ACP / ANP / MCP) — API Activity (6003).

    A single generic event for inter-agent messaging across protocols. These
    protocols are RPC over HTTP, so the message is an API operation; the wire
    protocol is a discriminator on the ``agent_message`` object rather than a
    dedicated class per protocol.

    Previously emitted as provisional ``agent_communication`` (9003) in a
    proposed ``ai`` category that OCSF never ratified. ``agent_message`` itself
    remains an AITF-proposed object — it did **not** ship in v1.9.0 — and is
    carried as an extension pending the agent-message PR.
    """
    category_uid: int = OCSFCategoryUID.APPLICATION
    class_uid: int = OCSFClassUID.API_ACTIVITY
    agent_message: OCSFAgentMessage


class AIToolExecutionEvent(AIBaseEvent):
    """Tool/MCP/function execution — reuses OCSF API Activity (6003)."""
    category_uid: int = OCSFCategoryUID.APPLICATION
    class_uid: int = OCSFClassUID.API_ACTIVITY
    tool_name: str
    tool_type: str  # "function", "mcp_tool", "skill", "api"
    tool_input: str | None = None
    tool_output: str | None = None
    is_error: bool = False
    duration_ms: float | None = None
    mcp_server: str | None = None
    mcp_transport: str | None = None
    skill_category: str | None = None
    skill_version: str | None = None
    approval_required: bool = False
    approved: bool | None = None


class AIDataRetrievalEvent(AIBaseEvent):
    """RAG / vector search — reuses OCSF Datastore Activity (6005)."""
    category_uid: int = OCSFCategoryUID.APPLICATION
    class_uid: int = OCSFClassUID.DATASTORE_ACTIVITY
    database_name: str
    database_type: str
    query: str | None = None
    top_k: int | None = None
    results_count: int = 0
    min_score: float | None = None
    max_score: float | None = None
    filter: str | None = None
    embedding_model: str | None = None
    embedding_dimensions: int | None = None
    pipeline_name: str | None = None
    pipeline_stage: str | None = None
    quality_scores: dict[str, float] | None = None


class AISecurityFindingEvent(AIBaseEvent):
    """AI security finding — reuses OCSF Detection Finding (2004)."""
    category_uid: int = OCSFCategoryUID.FINDINGS
    class_uid: int = OCSFClassUID.DETECTION_FINDING
    finding: AISecurityFinding


class AISupplyChainEvent(AIBaseEvent):
    """AI supply chain — reuses OCSF Vulnerability Finding (2002)."""
    category_uid: int = OCSFCategoryUID.FINDINGS
    class_uid: int = OCSFClassUID.VULNERABILITY_FINDING
    model_source: str
    model_hash: str | None = None
    model_license: str | None = None
    model_signed: bool = False
    model_signer: str | None = None
    verification_result: str | None = None  # "pass", "fail", "unknown"
    ai_bom_id: str | None = None
    ai_bom_components: str | None = None  # JSON
    ai_bom_format: str | None = None  # "aitf", "cyclonedx", "spdx"
    ai_bom_component_count: int | None = None
    ai_bom_vulnerability_count: int | None = None


class AIGovernanceEvent(AIBaseEvent):
    """AI governance/compliance — reuses OCSF Compliance Finding (2003)."""
    category_uid: int = OCSFCategoryUID.FINDINGS
    class_uid: int = OCSFClassUID.COMPLIANCE_FINDING
    frameworks: list[str] = Field(default_factory=list)
    controls: str | None = None  # JSON
    event_type: str = ""
    violation_detected: bool = False
    violation_severity: str | None = None
    remediation: str | None = None
    audit_id: str | None = None


class AIIdentityEvent(AIBaseEvent):
    """Agent identity/auth — reuses OCSF Authentication (3002, IAM).

    Delegation *lifecycle* maps to the new ``delegation_activity`` in the ai
    category; the delegation context itself rides on the ``ai_operation``
    profile (``delegation`` object) regardless of class.
    """
    category_uid: int = OCSFCategoryUID.IAM
    class_uid: int = OCSFClassUID.AUTHENTICATION
    agent_name: str
    agent_id: str
    auth_method: str  # "api_key", "oauth", "mtls", "jwt"
    auth_result: str  # "success", "failure", "denied"
    permissions: list[str] = Field(default_factory=list)
    credential_type: str | None = None
    delegation_chain: list[str] = Field(default_factory=list)
    scope: str | None = None


class AIModelOpsEvent(AIBaseEvent):
    """Model lifecycle ops — reuses OCSF Application Lifecycle (6002)."""
    category_uid: int = OCSFCategoryUID.APPLICATION
    class_uid: int = OCSFClassUID.APPLICATION_LIFECYCLE
    operation_type: str  # "training", "evaluation", "deployment", "serving", "monitoring", "prompt"
    model_id: str | None = None
    run_id: str | None = None
    framework: str | None = None
    status: str | None = None
    # Training-specific
    training_type: str | None = None
    base_model: str | None = None
    dataset_id: str | None = None
    epochs: int | None = None
    loss_final: float | None = None
    output_model_id: str | None = None
    # Evaluation-specific
    evaluation_type: str | None = None
    metrics: str | None = None  # JSON
    passed: bool | None = None
    # Deployment-specific
    deployment_id: str | None = None
    strategy: str | None = None
    environment: str | None = None
    endpoint: str | None = None
    # Serving-specific
    selected_model: str | None = None
    fallback_chain: str | None = None  # JSON
    cache_hit: bool | None = None
    # Monitoring-specific
    check_type: str | None = None
    drift_score: float | None = None
    drift_type: str | None = None
    action_triggered: str | None = None


class AIAssetInventoryEvent(AIBaseEvent):
    """AI asset inventory — reuses OCSF Inventory Info (5001, Discovery)."""
    category_uid: int = OCSFCategoryUID.DISCOVERY
    class_uid: int = OCSFClassUID.INVENTORY_INFO
    operation_type: str  # "register", "discover", "audit", "classify", "decommission"
    asset_id: str | None = None
    asset_name: str | None = None
    asset_type: str | None = None
    asset_version: str | None = None
    owner: str | None = None
    deployment_environment: str | None = None
    risk_classification: str | None = None
    # Discovery-specific
    discovery_scope: str | None = None
    discovery_method: str | None = None
    assets_found: int | None = None
    new_assets: int | None = None
    shadow_assets: int | None = None
    # Audit-specific
    audit_type: str | None = None
    audit_result: str | None = None
    audit_framework: str | None = None
    audit_findings: str | None = None  # JSON
    # Classification-specific
    classification_framework: str | None = None
    previous_classification: str | None = None
    classification_reason: str | None = None

File: telemetry/aitf/sdk/typescript/src/ocsf/event-classes.ts
/**
 * AITF OCSF AI Event Classes.
 *
 * AITF AI events reuse released OCSF classes rather than bespoke ones. OCSF
 * v1.9.0 attached the `ai_operation` profile to the `system`, `network`,
 * `application` and `iam` base classes, so every AI event has a released
 * home. Each factory sets both `category_uid` and `class_uid` to the reused
 * OCSF class; AI specifics ride on the `ai_operation` profile.
 * Based on event classes from the AITelemetry project, extended
 * for AITF with MCP, Skills, ModelOps, Asset Inventory, and enhanced agent support.
 */

import {
  AIBaseEvent,
  AICostInfo,
  AILatencyMetrics,
  AIModelInfo,
  AISecurityFinding,
  AITeamInfo,
  AITokenUsage,
  OCSFAgentMessage,
  OCSFCategoryUID,
  OCSFClassUID,
  createBaseEvent,
  createTokenUsage,
} from "./schema";

/** AI model inference — reuses OCSF API Activity (6003). */
export interface AIModelInferenceEvent extends AIBaseEvent {
  model: AIModelInfo;
  token_usage: AITokenUsage;
  latency?: AILatencyMetrics;
  request_content?: string;
  response_content?: string;
  streaming: boolean;
  tools_provided: number;
  finish_reason: string;
  cost?: AICostInfo;
  error?: Record<string, unknown>;
}

/** Agent lifecycle — new OCSF `agent_activity` in the ai category (9). */
export interface AIAgentActivityEvent extends AIBaseEvent {
  agent_name: string;
  agent_id: string;
  agent_type: string;
  framework?: string;
  session_id: string;
  step_type?: string;
  step_index?: number;
  thought?: string;
  action?: string;
  observation?: string;
  delegation_target?: string;
  team_info?: AITeamInfo;
}

/**
 * Agent-to-agent communication (A2A / ACP / ANP / MCP).
 *
 * A single generic event for inter-agent messaging across protocols, in the
 * proposed OCSF `ai` category. The wire protocol is a discriminator on the
 * `agent_message` object rather than a dedicated class per protocol.
 */
export interface AIAgentCommunicationEvent extends AIBaseEvent {
  agent_message: OCSFAgentMessage;
}

/** Tool/MCP/function execution — reuses OCSF API Activity (6003). */
export interface AIToolExecutionEvent extends AIBaseEvent {
  tool_name: string;
  tool_type: string; // "function", "mcp_tool", "skill", "api"
  tool_input?: string;
  tool_output?: string;
  is_error: boolean;
  duration_ms?: number;
  mcp_server?: string;
  mcp_transport?: string;
  skill_category?: string;
  skill_version?: string;
  approval_required: boolean;
  approved?: boolean;
}

/** RAG / vector search — reuses OCSF Datastore Activity (6005). */
export interface AIDataRetrievalEvent extends AIBaseEvent {
  database_name: string;
  database_type: string;
  query?: string;
  top_k?: number;
  results_count: number;
  min_score?: number;
  max_score?: number;
  filter?: string;
  embedding_model?: string;
  embedding_dimensions?: number;
  pipeline_name?: string;
  pipeline_stage?: string;
  quality_scores?: Record<string, number>;
}

/** AI security finding — reuses OCSF Detection Finding (2004). */
export interface AISecurityFindingEvent extends AIBaseEvent {
  finding: AISecurityFinding;
}

/** AI supply chain — reuses OCSF Vulnerability Finding (2002). */
export interface AISupplyChainEvent extends AIBaseEvent {
  model_source: string;
  model_hash?: string;
  model_license?: string;
  model_signed: boolean;
  model_signer?: string;
  verification_result?: string; // "pass", "fail", "unknown"
  ai_bom_id?: string;
  ai_bom_components?: string; // JSON
}

/** AI governance/compliance — reuses OCSF Compliance Finding (2003). */
export interface AIGovernanceEvent extends AIBaseEvent {
  frameworks: string[];
  controls?: string; // JSON
  event_type: string;
  violation_detected: boolean;
  violation_severity?: string;
  remediation?: string;
  audit_id?: string;
}

/** Agent identity/auth — reuses OCSF Authentication (3002, IAM). */
export interface AIIdentityEvent extends AIBaseEvent {
  agent_name: string;
  agent_id: string;
  auth_method: string; // "api_key", "oauth", "mtls", "jwt"
  auth_result: string; // "success", "failure", "denied"
  permissions: string[];
  credential_type?: string;
  delegation_chain: string[];
  scope?: string;
}

/** Model lifecycle ops — reuses OCSF Application Lifecycle (6002). */
export interface AIModelOpsEvent extends AIBaseEvent {
  operation_type: string; // "training", "evaluation", "deployment", "serving", "monitoring", "prompt"
  model_id?: string;
  run_id?: string;
  framework?: string;
  status?: string;
  training_type?: string;
  base_model?: string;
  dataset_id?: string;
  epochs?: number;
  loss_final?: number;
  output_model_id?: string;
  evaluation_type?: string;
  metrics?: string; // JSON
  passed?: boolean;
  deployment_id?: string;
  strategy?: string;
  environment?: string;
  endpoint?: string;
  selected_model?: string;
  fallback_chain?: string; // JSON
  cache_hit?: boolean;
  check_type?: string;
  drift_score?: number;
  drift_type?: string;
  action_triggered?: string;
}

/** AI asset inventory — reuses OCSF Inventory Info (5001, Discovery). */
export interface AIAssetInventoryEvent extends AIBaseEvent {
  operation_type: string; // "register", "discover", "audit", "classify", "decommission"
  asset_id?: string;
  asset_name?: string;
  asset_type?: string;
  asset_version?: string;
  owner?: string;
  deployment_environment?: string;
  risk_classification?: string;
  discovery_scope?: string;
  discovery_method?: string;
  assets_found?: number;
  new_assets?: number;
  shadow_assets?: number;
  audit_type?: string;
  audit_result?: string;
  audit_framework?: string;
  audit_findings?: string; // JSON
  classification_framework?: string;
  previous_classification?: string;
  classification_reason?: string;
}

// --- Validation Helpers ---

/**
 * Valid OCSF (class_uid -> category_uid) pairs that AITF reuses.
 *
 * AITF reuses existing OCSF classes (OCSF PR #1641 / issue #1640); each class
 * lives under its own OCSF category, so class_uid is no longer unique per AITF
 * event type (inference and tool execution both reuse API Activity 6003).
 */
const VALID_CLASS_CATEGORY: ReadonlyMap<number, number> = new Map([
  [OCSFClassUID.API_ACTIVITY, OCSFCategoryUID.APPLICATION],
  [OCSFClassUID.DATASTORE_ACTIVITY, OCSFCategoryUID.APPLICATION],
  [OCSFClassUID.APPLICATION_LIFECYCLE, OCSFCategoryUID.APPLICATION],
  [OCSFClassUID.DETECTION_FINDING, OCSFCategoryUID.FINDINGS],
  [OCSFClassUID.VULNERABILITY_FINDING, OCSFCategoryUID.FINDINGS],
  [OCSFClassUID.COMPLIANCE_FINDING, OCSFCategoryUID.FINDINGS],
  [OCSFClassUID.AUTHENTICATION, OCSFCategoryUID.IAM],
  [OCSFClassUID.INVENTORY_INFO, OCSFCategoryUID.DISCOVERY],
  [OCSFClassUID.AUTHORIZE_SESSION, OCSFCategoryUID.IAM],
]);

/**
 * Validate that the class_uid and category_uid form a known reused OCSF pair.
 * Called internally by factory functions after creating the base event.
 */
function validateBaseFields(event: AIBaseEvent, factoryName: string): void {
  const expectedCategory = VALID_CLASS_CATEGORY.get(event.class_uid);
  if (expectedCategory === undefined) {
    throw new Error(
      `${factoryName}: invalid class_uid ${event.class_uid}. ` +
        `Expected one of: ${[...VALID_CLASS_CATEGORY.keys()].join(", ")}`
    );
  }
  if (event.category_uid !== expectedCategory) {
    throw new Error(
      `${factoryName}: invalid category_uid ${event.category_uid} for ` +
        `class_uid ${event.class_uid}. Expected ${expectedCategory}.`
    );
  }
}

/**
 * Validate that a required string field is present and non-empty.
 */
function requireString(
  value: unknown,
  fieldName: string,
  factoryName: string
): void {
  if (typeof value !== "string" || value.length === 0) {
    throw new Error(
      `${factoryName}: required field '${fieldName}' must be a non-empty string.`
    );
  }
}

// --- Factory Functions ---

/** Create a Model Inference event. */
export function createModelInferenceEvent(
  options: {
    model: AIModelInfo;
    tokenUsage?: Partial<AITokenUsage>;
    latency?: AILatencyMetrics;
    cost?: AICostInfo;
    streaming?: boolean;
    toolsProvided?: number;
    finishReason?: string;
    requestContent?: string;
    responseContent?: string;
    activityId?: number;
    message?: string;
    time?: string;
  }
): AIModelInferenceEvent {
  if (!options.model || typeof options.model.model_id !== "string" || options.model.model_id.length === 0) {
    throw new Error("createModelInferenceEvent: required field 'model.model_id' must be a non-empty string.");
  }
  const base = createBaseEvent(OCSFClassUID.API_ACTIVITY, {
    category_uid: OCSFCategoryUID.APPLICATION,
    activity_id: options.activityId,
    message: options.message,
    time: options.time,
  });
  validateBaseFields(base, "createModelInferenceEvent");

  return {
    ...base,
    model: options.model,
    token_usage: createTokenUsage(options.tokenUsage),
    latency: options.latency,
    request_content: options.requestContent,
    response_content: options.responseContent,
    streaming: options.streaming ?? false,
    tools_provided: options.toolsProvided ?? 0,
    finish_reason: options.finishReason ?? "stop",
    cost: options.cost,
  };
}

/**
 * Create an Agent Activity event on **API Activity (6003)**.
 *
 * AITF <= 0.4 emitted this as `agent_activity` (9001) in a proposed `ai`
 * category. That category was never ratified, so agent lifecycle now rides
 * API Activity with the `ai_operation` profile carrying `ai_agent`. Agent
 * session start/stop MAY instead use Application Lifecycle (6002) where the
 * agent is modelled as a managed application.
 */
export function createAgentActivityEvent(
  options: {
    agentName: string;
    agentId: string;
    sessionId: string;
    agentType?: string;
    framework?: string;
    stepType?: string;
    stepIndex?: number;
    thought?: string;
    action?: string;
    observation?: string;
    delegationTarget?: string;
    teamInfo?: AITeamInfo;
    activityId?: number;
    message?: string;
    time?: string;
  }
): AIAgentActivityEvent {
  requireString(options.agentName, "agentName", "createAgentActivityEvent");
  requireString(options.agentId, "agentId", "createAgentActivityEvent");
  requireString(options.sessionId, "sessionId", "createAgentActivityEvent");
  const base = createBaseEvent(OCSFClassUID.API_ACTIVITY, {
    category_uid: OCSFCategoryUID.APPLICATION,
    activity_id: options.activityId,
    message: options.message,
    time: options.time,
  });
  validateBaseFields(base, "createAgentActivityEvent");

  return {
    ...base,
    agent_name: options.agentName,
    agent_id: options.agentId,
    agent_type: options.agentType ?? "autonomous",
    framework: options.framework,
    session_id: options.sessionId,
    step_type: options.stepType,
    step_index: options.stepIndex,
    thought: options.thought,
    action: options.action,
    observation: options.observation,
    delegation_target: options.delegationTarget,
    team_info: options.teamInfo,
  };
}

/**
 * Create an Agent Communication event on **API Activity (6003)**.
 *
 * AITF <= 0.4 emitted this as `agent_communication` (9003). A2A, ACP and ANP
 * exchanges are request/response API calls between agents, so API Activity is
 * their released home; the protocol discriminator rides `agent_message`.
 */
export function createAgentCommunicationEvent(
  options: {
    agentMessage: OCSFAgentMessage;
    activityId?: number;
    statusId?: number;
    message?: string;
    time?: string;
  }
): AIAgentCommunicationEvent {
  if (!options.agentMessage || typeof options.agentMessage !== "object") {
    throw new Error(
      "createAgentCommunicationEvent: required field 'agentMessage' must be a valid OCSFAgentMessage object."
    );
  }
  const base = createBaseEvent(OCSFClassUID.API_ACTIVITY, {
    category_uid: OCSFCategoryUID.APPLICATION,
    activity_id: options.activityId,
    status_id: options.statusId,
    message: options.message,
    time: options.time,
  });
  validateBaseFields(base, "createAgentCommunicationEvent");

  return {
    ...base,
    agent_message: options.agentMessage,
  };
}

/** Create a Tool Execution event. */
export function createToolExecutionEvent(
  options: {
    toolName: string;
    toolType: string;
    toolInput?: string;
    toolOutput?: string;
    isError?: boolean;
    durationMs?: number;
    mcpServer?: string;
    mcpTransport?: string;
    skillCategory?: string;
    skillVersion?: string;
    approvalRequired?: boolean;
    approved?: boolean;
    activityId?: number;
    message?: string;
    time?: string;
  }
): AIToolExecutionEvent {
  requireString(options.toolName, "toolName", "createToolExecutionEvent");
  requireString(options.toolType, "toolType", "createToolExecutionEvent");
  const base = createBaseEvent(OCSFClassUID.API_ACTIVITY, {
    category_uid: OCSFCategoryUID.APPLICATION,
    activity_id: options.activityId,
    message: options.message,
    time: options.time,
  });
  validateBaseFields(base, "createToolExecutionEvent");

  return {
    ...base,
    tool_name: options.toolName,
    tool_type: options.toolType,
    tool_input: options.toolInput,
    tool_output: options.toolOutput,
    is_error: options.isError ?? false,
    duration_ms: options.durationMs,
    mcp_server: options.mcpServer,
    mcp_transport: options.mcpTransport,
    skill_category: options.skillCategory,
    skill_version: options.skillVersion,
    approval_required: options.approvalRequired ?? false,
    approved: options.approved,
  };
}

/** Create a Data Retrieval event. */
export function createDataRetrievalEvent(
  options: {
    databaseName: string;
    databaseType: string;
    query?: string;
    topK?: number;
    resultsCount?: number;
    minScore?: number;
    maxScore?: number;
    filter?: string;
    embeddingModel?: string;
    embeddingDimensions?: number;
    pipelineName?: string;
    pipelineStage?: string;
    qualityScores?: Record<string, number>;
    activityId?: number;
    message?: string;
    time?: string;
  }
): AIDataRetrievalEvent {
  requireString(options.databaseName, "databaseName", "createDataRetrievalEvent");
  requireString(options.databaseType, "databaseType", "createDataRetrievalEvent");
  const base = createBaseEvent(OCSFClassUID.DATASTORE_ACTIVITY, {
    category_uid: OCSFCategoryUID.APPLICATION,
    activity_id: options.activityId,
    message: options.message,
    time: options.time,
  });
  validateBaseFields(base, "createDataRetrievalEvent");

  return {
    ...base,
    database_name: options.databaseName,
    database_type: options.databaseType,
    query: options.query,
    top_k: options.topK,
    results_count: options.resultsCount ?? 0,
    min_score: options.minScore,
    max_score: options.maxScore,
    filter: options.filter,
    embedding_model: options.embeddingModel,
    embedding_dimensions: options.embeddingDimensions,
    pipeline_name: options.pipelineName,
    pipeline_stage: options.pipelineStage,
    quality_scores: options.qualityScores,
  };
}

/** Create a Security Finding event. */
export function createSecurityFindingEvent(
  options: {
    finding: AISecurityFinding;
    severityId?: number;
    activityId?: number;
    message?: string;
    time?: string;
  }
): AISecurityFindingEvent {
  if (!options.finding || typeof options.finding !== "object") {
    throw new Error("createSecurityFindingEvent: required field 'finding' must be a valid AISecurityFinding object.");
  }
  requireString(options.finding.finding_type, "finding.finding_type", "createSecurityFindingEvent");
  requireString(options.finding.risk_level, "finding.risk_level", "createSecurityFindingEvent");
  const base = createBaseEvent(OCSFClassUID.DETECTION_FINDING, {
    category_uid: OCSFCategoryUID.FINDINGS,
    activity_id: options.activityId ?? 1,
    severity_id: options.severityId,
    message: options.message,
    time: options.time,
  });
  validateBaseFields(base, "createSecurityFindingEvent");

  return {
    ...base,
    finding: options.finding,
  };
}

/** Create a Supply Chain event. */
export function createSupplyChainEvent(
  options: {
    modelSource: string;
    modelHash?: string;
    modelLicense?: string;
    modelSigned?: boolean;
    modelSigner?: string;
    verificationResult?: string;
    aiBomId?: string;
    aiBomComponents?: string;
    activityId?: number;
    message?: string;
    time?: string;
  }
): AISupplyChainEvent {
  requireString(options.modelSource, "modelSource", "createSupplyChainEvent");
  const base = createBaseEvent(OCSFClassUID.VULNERABILITY_FINDING, {
    category_uid: OCSFCategoryUID.FINDINGS,
    activity_id: options.activityId,
    message: options.message,
    time: options.time,
  });
  validateBaseFields(base, "createSupplyChainEvent");

  return {
    ...base,
    model_source: options.modelSource,
    model_hash: options.modelHash,
    model_license: options.modelLicense,
    model_signed: options.modelSigned ?? false,
    model_signer: options.modelSigner,
    verification_result: options.verificationResult,
    ai_bom_id: options.aiBomId,
    ai_bom_components: options.aiBomComponents,
  };
}

/** Create a Governance event. */
export function createGovernanceEvent(
  options: {
    frameworks?: string[];
    controls?: string;
    eventType?: string;
    violationDetected?: boolean;
    violationSeverity?: string;
    remediation?: string;
    auditId?: string;
    activityId?: number;
    message?: string;
    time?: string;
  }
): AIGovernanceEvent {
  const base = createBaseEvent(OCSFClassUID.COMPLIANCE_FINDING, {
    category_uid: OCSFCategoryUID.FINDINGS,
    activity_id: options.activityId,
    message: options.message,
    time: options.time,
  });
  validateBaseFields(base, "createGovernanceEvent");

  return {
    ...base,
    frameworks: options.frameworks ?? [],
    controls: options.controls,
    event_type: options.eventType ?? "",
    violation_detected: options.violationDetected ?? false,
    violation_severity: options.violationSeverity,
    remediation: options.remediation,
    audit_id: options.auditId,
  };
}

/** Create an Identity event. */
export function createIdentityEvent(
  options: {
    agentName: string;
    agentId: string;
    authMethod: string;
    authResult: string;
    permissions?: string[];
    credentialType?: string;
    delegationChain?: string[];
    scope?: string;
    activityId?: number;
    message?: string;
    time?: string;
  }
): AIIdentityEvent {
  requireString(options.agentName, "agentName", "createIdentityEvent");
  requireString(options.agentId, "agentId", "createIdentityEvent");
  requireString(options.authMethod, "authMethod", "createIdentityEvent");
  requireString(options.authResult, "authResult", "createIdentityEvent");
  const base = createBaseEvent(OCSFClassUID.AUTHENTICATION, {
    category_uid: OCSFCategoryUID.IAM,
    activity_id: options.activityId,
    message: options.message,
    time: options.time,
  });
  validateBaseFields(base, "createIdentityEvent");

  return {
    ...base,
    agent_name: options.agentName,
    agent_id: options.agentId,
    auth_method: options.authMethod,
    auth_result: options.authResult,
    permissions: options.permissions ?? [],
    credential_type: options.credentialType,
    delegation_chain: options.delegationChain ?? [],
    scope: options.scope,
  };
}

/** Create a Model Operations event. */
export function createModelOpsEvent(
  options: {
    operationType: string;
    modelId?: string;
    runId?: string;
    framework?: string;
    status?: string;
    trainingType?: string;
    baseModel?: string;
    datasetId?: string;
    epochs?: number;
    lossFinal?: number;
    outputModelId?: string;
    evaluationType?: string;
    metrics?: string;
    passed?: boolean;
    deploymentId?: string;
    strategy?: string;
    environment?: string;
    endpoint?: string;
    selectedModel?: string;
    fallbackChain?: string;
    cacheHit?: boolean;
    checkType?: string;
    driftScore?: number;
    driftType?: string;
    actionTriggered?: string;
    activityId?: number;
    message?: string;
    time?: string;
  }
): AIModelOpsEvent {
  requireString(options.operationType, "operationType", "createModelOpsEvent");
  const base = createBaseEvent(OCSFClassUID.APPLICATION_LIFECYCLE, {
    category_uid: OCSFCategoryUID.APPLICATION,
    activity_id: options.activityId,
    message: options.message,
    time: options.time,
  });
  validateBaseFields(base, "createModelOpsEvent");

  return {
    ...base,
    operation_type: options.operationType,
    model_id: options.modelId,
    run_id: options.runId,
    framework: options.framework,
    status: options.status,
    training_type: options.trainingType,
    base_model: options.baseModel,
    dataset_id: options.datasetId,
    epochs: options.epochs,
    loss_final: options.lossFinal,
    output_model_id: options.outputModelId,
    evaluation_type: options.evaluationType,
    metrics: options.metrics,
    passed: options.passed,
    deployment_id: options.deploymentId,
    strategy: options.strategy,
    environment: options.environment,
    endpoint: options.endpoint,
    selected_model: options.selectedModel,
    fallback_chain: options.fallbackChain,
    cache_hit: options.cacheHit,
    check_type: options.checkType,
    drift_score: options.driftScore,
    drift_type: options.driftType,
    action_triggered: options.actionTriggered,
  };
}

/** Create an Asset Inventory event. */
export function createAssetInventoryEvent(
  options: {
    operationType: string;
    assetId?: string;
    assetName?: string;
    assetType?: string;
    assetVersion?: string;
    owner?: string;
    deploymentEnvironment?: string;
    riskClassification?: string;
    discoveryScope?: string;
    discoveryMethod?: string;
    assetsFound?: number;
    newAssets?: number;
    shadowAssets?: number;
    auditType?: string;
    auditResult?: string;
    auditFramework?: string;
    auditFindings?: string;
    classificationFramework?: string;
    previousClassification?: string;
    classificationReason?: string;
    activityId?: number;
    message?: string;
    time?: string;
  }
): AIAssetInventoryEvent {
  requireString(options.operationType, "operationType", "createAssetInventoryEvent");
  const base = createBaseEvent(OCSFClassUID.INVENTORY_INFO, {
    category_uid: OCSFCategoryUID.DISCOVERY,
    activity_id: options.activityId,
    message: options.message,
    time: options.time,
  });
  validateBaseFields(base, "createAssetInventoryEvent");

  return {
    ...base,
    operation_type: options.operationType,
    asset_id: options.assetId,
    asset_name: options.assetName,
    asset_type: options.assetType,
    asset_version: options.assetVersion,
    owner: options.owner,
    deployment_environment: options.deploymentEnvironment,
    risk_classification: options.riskClassification,
    discovery_scope: options.discoveryScope,
    discovery_method: options.discoveryMethod,
    assets_found: options.assetsFound,
    new_assets: options.newAssets,
    shadow_assets: options.shadowAssets,
    audit_type: options.auditType,
    audit_result: options.auditResult,
    audit_framework: options.auditFramework,
    audit_findings: options.auditFindings,
    classification_framework: options.classificationFramework,
    previous_classification: options.previousClassification,
    classification_reason: options.classificationReason,
  };
}
